File: src/karkinolution/systems/death.py
from karkinolution.organism.creatures import (
    Corpse,
    Creature,
    PregnantUterus,
)
from karkinolution.organism.identity import gen_id
from karkinolution.organism.stats import (
    Age,
    Energy,
)
            
        
class DeathSystem:
    @staticmethod
    def generate_corpse(creature:Creature) -> Corpse:
        energy:float = creature.energy.value * 0.8
        if creature.pregnant:
            assert isinstance(creature.uterus, PregnantUterus)
            energy += creature.uterus.pregnancy_cost * 2

        decomposition_time = round(energy/creature.energy.limit * 7.5)
        return Corpse(Energy(energy, energy), gen_id(), Age(0, decomposition_time), creature.position)
    @staticmethod
    def is_dead(creature:Creature) -> bool:
        if creature.life.value == 0:
            return True
        if creature.energy.value == 0:
            return True
        if creature.age.value >= creature.age.limit:
            return True
        return False
          

File: src/karkinolution/systems/death.cpp
#include <karkinolution/organism/genetics.hpp>
#include <karkinolution/organism/ontology.hpp>
#include <karkinolution/core/stats.hpp>
#include <karkinolution/organism/creatures.hpp>
#include <karkinolution/systems/death.hpp>
#include <cmath>

Corpse DeathSystem::generate_corpse(const Creature&creature) {
    float energy = creature.energy.value() * 0.8;
    if (creature.pregnant()) {
        energy += std::get<PregnantUterus>(creature.uterus).pregnancy_cost() * 2;
    }
    IntegerLimited<uint16_t> decomposition_time = {std::round(energy/creature.energy.max() * 7.5)};
    return Corpse{
        Energy(energy, energy),
        IDF::gen_id(),
        decomposition_time,
        creature.position
    };
}

bool DeathSystem::is_dead(const Creature &creature) {
    return creature.life.value() < 1e-9 || creature.energy.value() < 1e-9 || creature.age.value() >= creature.age.max();
}
